ml-agents/mlagents/trainers/trainer_controller.py

```diff
@@ -181,24 +183,34 @@ def write_to_tensorboard(self, global_step: int) -> None:
             else:
                 trainer.write_summary(global_step, delta_train_start)
 
+    def start_trainer(self, trainer: Trainer, env_manager: EnvManager) -> None:
+        self.trainers[trainer.brain_name] = trainer
+        self.logger.info(trainer)
+        if self.train_model:
+            trainer.write_tensorboard_text("Hyperparameters", trainer.parameters)
+        env_manager.set_policy(trainer.brain_name, trainer.policy)
+
     def start_learning(self, env_manager: EnvManager) -> None:
         self._create_model_path(self.model_path)
 
         tf.reset_default_graph()
 
-        for _, t in self.trainers.items():
-            self.logger.info(t)
-
         global_step = 0
 
-        if self.train_model:
-            for brain_name, trainer in self.trainers.items():
-                trainer.write_tensorboard_text("Hyperparameters", trainer.parameters)
         try:
-            for brain_name, trainer in self.trainers.items():
-                env_manager.set_policy(brain_name, trainer.policy)
             self._reset_env(env_manager)
             while self._not_done_training():
+                if self.last_brain_names != env_manager.external_brains.keys():
```

Comment:
`last_brain_names` doesn't need to be a class member, since its state is only local to `start_learning`.  Could you move its initialization into `start_learning`?